File: 1762-furthest-building-you-can-reach/solution.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int len=heights.size();
        priority_queue<int, vector<int>, greater<int> > pq;
        int idx=0;
        int ct=0;
        for(int i=1;i<len;i++){
            int dif=heights[i]-heights[i-1];
            if(dif>0){
                if(ct<ladders){
                    pq.push(dif);
                    idx=i;
                    ct++;
                }else{
                    int mn=dif;
                    if(!pq.empty())mn=min(pq.top(),mn);
                    if(mn<=bricks){
                        pq.push(dif);
                        pq.pop();
                        bricks-=mn;
                        idx=i;
                    }else break;
                }
            }else idx=i;
        }
        return idx;
    }
};

File: 1762-furthest-building-you-can-reach/solution.cs
public class Solution {
    public int FurthestBuilding(int[] heights, int bricks, int ladders) {

        PriorityQueue<int,int> pq = new PriorityQueue<int,int>();

        int ans = heights.Length - 1;

        for(int i = 1; i < heights.Length; i++){

            int diff = heights[i] - heights[i-1];

            if(diff > 0){
                bricks -= diff;
                pq.Enqueue(diff, -diff);
            }

            if(bricks < 0){

                if(ladders > 0){
                    
                    bricks += pq.Dequeue();
                    ladders--;
                }else
                {
                    ans = i - 1;
                    break;
                }
            }
        }

        return ans;
    }
}
